---
title: "matrix processing"
author: "Zhuoer Dong"
date: "November 6, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
```



```{r message=FALSE}
```

# To Do
- 使用哪两类/多类样本
- 路径
- filter写成函数 外界传入要求
- top k 阈值
- 串起来时变量冲突问题
- batch removal需要log再变回去
- cpm，remove pi、mi，top20，reference gene等方法并入R
- RUVs 用或者不用样本信息
- combat必须用类别信息
- alignment score
- ruvs, combat log and reverse
- scnorm分析图
- ruvs分析图
- 确保输入数据格式没有问题

filename, savename, use which operation, pipeline
- Imputation
    - scImpute
- Normalization
    - CPM
    - Top20
    - Remove pi, miRNA
    - TMM
    - RLE
    - SCNorm
- Batch Removal
    - RUVs: need logged normalization matrix
    - Combat: need logged normalization matrix

# load data


```{r read-data}
matrix_path = 'scirep_sequential_qc.txt'
classinfo_path = 'scirep_classes.txt'


mat <- matrix_path %>% rexseek::read_mat()
info <- classinfo_path %>% readr::read_csv() %T>% print
```

```{r filter-low-value}
thres_count = 2
thres_sample_nums = 5

df <- df_raw[rowSums(df_raw > thres_count) > thres_sample_nums, ] %T>% print(n_extra = 0)
```

```{r to-matrix}
mat <- df %>% as.data.frame() %>% 
    tibble::column_to_rownames('transcript') %>% as.matrix()
```

```{r make-SingleCellExperiment}
reads <- SingleCellExperiment::SingleCellExperiment(
    assays =  list(counts = mat), colData = info
) %T>% print
```


```{r plotHighestExprs}
reads %>% scater::calculateQCMetrics() %>% scater::plotHighestExprs(n = 20)
```

```{r plotTSNE}
reads %>% scater::plotTSNE(
    shape_by = "label", colour_by = "label", 
    run_args = list(exprs_values = "counts")
)
```


```{r imputation, eval=F}
impute_tmpdir <- tempdir();
impute_infile <- paste0(impute_tmpdir, '/pre_counts.csv');
impute_outdir <- paste0(impute_tmpdir, '/output')

mat %T>% {colnames(.)[1] <- ''} %>% readr::write_csv(impute_infile)

#callr::r(
do.call(
    function(...) {scImpute::scimpute(...)}, 
    args = list(
        'tmpsave.csv', out_dir = impute_outdir, Kcluster = 5, ncores = 2, labeled = TRUE, 
        labels = colnames(mat) %>% {.[-1]} %>% plyr::mapvalues(info[[1]], info[[2]], F)
    )
)

dir(impute_outdir)
```





```{r SCnorm}
NormalizedData <- SCnorm::SCnorm(mat, rep(1, ncol(mat)), T, NCores = 8) %>% 
    SCnorm::results()
```



```{r TMM}
## normalize the object using the saved size factors
reads_tmm <- scater::normaliseExprs(reads, "TMM") %>% 
    scater::normalize()
```



```{r RLE}
## normalize the object using the saved size factors
reads_rle <- scater::normaliseExprs(reads, "RLE") %>%
    scater::normalize()
```



```{r}
```


