shell.prefix('set -x;')
configfile: 'snakemake/config.json'

def get_all_inputs(wildcards):
    available_inputs = dict(
        fasta_index=expand('data/fasta/{rna_type}.fa.fai', rna_type=rna_types)
    )
    enabled_inputs = list(available_inputs.keys())
    inputs = []
    for key, l in available_inputs.items():
        if key in enabled_inputs:
            inputs += l
    return inputs

rule all:
    input:
        get_all_inputs

rule fasta_index:
    input:
        '{genome_dir}/fasta/{rna_type}.fa'
    output:
        '{genome_dir}/fasta/{rna_type}.fa.fai'
    shell:
        '''samtools faidx {input}'''

rule gtf_to_bed12:
    input:
        '{genome_dir}/gtf/{rna_type}.gtf'
    output:
        '{genome_dir}/bed12/{rna_type}.bed'
    shell:
        '''gffread --bed -o {output} {input}
        '''

rule fasta_index_to_tbed:
    input:
        '{genome_dir}/fasta/{rna_type}.fa.fai'
    output:
        '{genome_dir}/tbed/{rna_type}.bed'
    shell:
        '''awk 'BEGIN{{OFS="\t"}}{{print $1,0,$2,$1,0,"+"}}' {input} | bedtools sort > {output}
        '''

rule fasta_index_to_chrom_sizes:
    input:
        '{genome_dir}/fasta/{rna_type}.fa.fai'
    output:
        '{genome_dir}/chrom_sizes/{rna_type}'
    shell:
        '''cut -d'\t' -f1,2 {input} > {output}
        '''